### PR TITLE
[MRG] Refactor test_func_inspect.py as per pytest design.

### PR DESCRIPTION
**Pull request by karandesai-96**
#### Third Phase PR on #411 (Succeeding PR #446)

This PR deals with refactor of tests in **test_func_inspect.py** to adopt the pytest flavor.

* Prior to this PR, each method had several `yield` based statements where the test cases comprised of a set of functions with args and kwargs as well. I have proposed a `fixture` which returns a dictionary of functions to make `parametrization` easier.

* `with_numpy` decorator is now a `pytest.skipif` marker [Commented for now, will be uncommented after **test_hashing.py** is pytestified. Because this skipif marker does not work well with `yield` based tests.]

* The function names `g` and `f2` have been interchanged. Earlier `g` was decorated with `@mem.cache` earlier, and it was the only one function. All other functions - `f`, `f2`, `h`, `i`, `j`, `k` were normal undecorated ones. This name interchanging is simply a cosmetic change.

### CODE REVIEW

**Comment by karandesai-96** [comment]
@lesteve It is WIP but since the diff is small, it might be easier for review, feel free to review it whenever you wish. :smile:

File: joblib/test/test_func_inspect.py [hunk 1]
```diff
@@ -81,58 +62,61 @@ def f(self, x):
...
-    yield assert_equal, filter_args(f, ['y'], (0, )), {'x': 0}
-    yield assert_equal, filter_args(f, ['y'], (0, ), dict(y=1)), {'x': 0}
-    yield assert_equal, filter_args(f, ['x', 'y'], (0, )), {}
-    yield assert_equal, filter_args(f, [], (0,), dict(y=1)), {'x': 0, 'y': 1}
-    yield assert_equal, filter_args(f, ['y'], (), dict(x=2, y=1)), {'x': 2}
 
-    yield assert_equal, filter_args(i, [], (2, )), {'x': 2}
-    yield assert_equal, filter_args(f2, [], (), dict(x=1)), {'x': 1}
+@parametrize(['funcname', 'args', 'filtered_args'],
+             [('f', [[], (1, )], {'x': 1, 'y': 0}),
```
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: Why not use functions in parametrize?

```py
@parametrize(['func', 'args', 'filtered_args'],
    [(f, [[], (1, )], {'x': 1, 'y': 0}),
    ...
```

I guess the only good reason to use a fixture for funcs is actually the cached function `f2`, because it depends on tmpdir, right?

Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: Yes, because `f2` depends on tmpdir. So would it be great if `f2` is enclosed in a fixture while other functions just stay there in global scope ?

**Comment by karandesai-96** [comment]
I am rolling back to old `with_numpy` because the `skipif` marker will not work with `yield` based tests yet - as in in failing job of Travis CI. But only four such tests exist in **test_hashing.py** so I will bring this back in immediately next PR.

File: joblib/test/test_func_inspect.py [hunk 2]
```diff
@@ -6,54 +6,24 @@
...
-            shutil.rmtree(temp_folder)
-        except Exception as e:
-            print("Failed to delete temporary folder %s: %r" %
-                  (temp_folder, e))
-
-
-@mem.cache
 def g(x):
-    return x
+    pass
```
Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: I interchanged the names of `g`g and `f2` as `f2` was basically the "odd one out..."

**Comment by karandesai-96** [comment]
@lesteve Anything more that should go in here?

File: joblib/test/test_func_inspect.py (hunk 2 above)
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: Why this change?

File: joblib/test/test_func_inspect.py [hunk 3]
```diff
@@ -6,54 +6,24 @@
...
-import shutil
-import tempfile
 import functools
 
 from joblib.func_inspect import filter_args, get_func_name, get_func_code
 from joblib.func_inspect import _clean_win_chars, format_signature
 from joblib.memory import Memory
 from joblib.test.common import with_numpy
-from joblib.testing import assert_equal, assert_raises_regex, assert_raises
+from joblib.testing import (fixture, parametrize, pytest_assert_raises)
```
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: You don't need the parentheses if the imports fit on one line

File: joblib/test/test_func_inspect.py [hunk 4]
```diff
@@ -81,58 +70,69 @@ def f(self, x):
...
-    yield assert_equal, filter_args(h, [], (1, 25), dict(ee=2)), \
-                            {'x': 1, 'y': 25, '*': [], '**': {'ee': 2}}
-    yield assert_equal, filter_args(h, ['*'], (1, 2, 25), dict(ee=2)), \
-                            {'x': 1, 'y': 2, '**': {'ee': 2}}
+@parametrize(['func', 'args', 'filtered_args'],
+             [(h, [[], (1, )],
+               {'x': 1, 'y': 0, '*': [], '**': {}}),
+              (h, [[], (1, 2, 3, 4)],
+               {'x': 1, 'y': 2, '*': [3, 4], '**': {}}),
+              (h, [[], (1, 25), dict(ee=2)],
```
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: For consistency you may as well use {'ee': 2}. To be honest I am not sure why `dict` is used maybe some people find it more explicit.

If you do that do it everywhere in this file.

File: joblib/test/test_func_inspect.py [hunk 5]
```diff
@@ -72,6 +42,25 @@ def k(*args, **kwargs):
     pass
 
 
+@fixture(scope='module')
+def f2(tmpdir_factory):
+    """
+    This fixture returns a dict of functions which are used as test cases
```
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: This docstring is lying.

File: joblib/test/test_func_inspect.py [hunk 6]
```diff
@@ -81,58 +70,69 @@ def f(self, x):
...
-    yield assert_equal, 'g', get_func_name(g)[1]
+@parametrize(['func', 'funcname'], [(f, 'f'), (g, 'g'), (f2, 'f2')])
+def test_func_name(func, funcname):
+    # Check that we are not confused by decoration
+    # here testcase 'f2' is the function itself
+    assert get_func_name(func)[1] == funcname
+
+
+def test_func_name_on_inner_func(f2):
+    # Check that we are not confused by decoration
```
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: I don't understand why the "not confused" test was on `g` before and is on `f2` now.

File: joblib/test/test_func_inspect.py (hunk 2 above)
Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: I interchanged names `g` and `f2` just because `f2` was the odd one out. Maybe it got missed out.

File: joblib/test/test_func_inspect.py (hunk 6 above)
Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: `g` was decorated with `@mem.cache` earlier, and it was the only one function. All other functions - `f`, `f2`, `h`, `i`, `j`, `k` were normal ones. That seemed a bit off to me to I went ahead and made this change.

File: joblib/test/test_func_inspect.py (hunk 2 above)
Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: Nope, this one is correct. Earlier it was:

```python
def f2(x):
    pass

@mem.cache
def g(x):
    return x
```

Just the names `g` and `f2` were interchanged because `f2` was the odd one out (decorated).

File: joblib/test/test_func_inspect.py (hunk 6 above)
Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: I added this note in main PR description.

**Comment by karandesai-96** [comment]
@lesteve All work done here I guess. New PR on the way once this is merged..

File: joblib/test/test_func_inspect.py [hunk 7]
```diff
@@ -72,6 +42,21 @@ def k(*args, **kwargs):
     pass
 
 
+@fixture(scope='module')
+def f2(tmpdir_factory):
+    # Create a Memory object to test decorated functions.
+    # We should be careful not to call the decorated functions, so that
+    # cache directories are not created in the temp dir.
```
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: I have to say I don't really understand this comment but it was there before ...

Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: Agreed. I did not alter it because it was there...

File: joblib/test/test_func_inspect.py [hunk 8]
```diff
@@ -193,11 +188,22 @@ def test_bound_methods():
...
+             [(ValueError, 'ignore_lst must be a list of parameters to ignore',
+               f, ['bar', (None, )]),
+              (ValueError, 'Ignore list: argument \'(.*)\' is not defined',
+               g, [['bar'], (None, )]),
+              (ValueError, 'Wrong number of arguments',
+               h, [[]])
+              ])
+def test_filter_args_error_msg(exception, regex, func, args):
+    """
+    Make sure that filter_args returns decent error messages, for the sake
```
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: Was this docstring too long? It it was try to follow [PEP0258](https://www.python.org/dev/peps/pep-0258/). Short description one the first line and then a blank line and then long description.

File: joblib/test/test_func_inspect.py [hunk 9]
```diff
@@ -207,13 +213,13 @@ def test_clean_win_chars():
         assert char not in mangled_string
 
 
-def test_format_signature():
+@parametrize(['func', 'args', 'kwargs', 'sgn_expected'],
+             [(g, [list(range(5))], {}, 'g([0, 1, 2, 3, 4])'),
+              (k, [1, 2, (3, 4)], {'y': True}, 'k(1, 2, (3, 4), y=True)')])
```
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: Your changes are fine, but remember what I said above about not changing unrelated things if there is not a strong motivation behind it.

File: joblib/test/test_func_inspect.py (hunk 8 above)
Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: Oh okay I did not remember this PEP. I usually write them throughout the project in this manner so I wrote them here in this way unconsciously.. I'll rectify them in a jiffy.

Comment (lesteve) [inline joblib/test/test_func_inspect.py]: For the record, in general we try to follow the Numpy style for docstrings: see https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt.

File: joblib/test/test_func_inspect.py (hunk 9 above)
Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: It was just to keep things within 80 characters. I would remember about the fact, keeping in mind that there might be some piece of code which would be written a long time ago due to some reason. :smile:

File: joblib/test/test_func_inspect.py (hunk 8 above)
Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: Okay, thanks for informing the style guidelines. Will be helpful ! (_pins this link to browser_ ).

I modified the docstring..

File: joblib/test/test_func_inspect.py (hunk 2 above)
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: > Since you are changing it anyway, can you use a better name than f2 then to make it clearer than it is a cached function? Maybe something like cached_func?

I think this is the last thing remaining and then this will be good to merge.

File: joblib/test/test_func_inspect.py (hunk 7 above)
Comment (lesteve) [inline joblib/test/test_func_inspect.py]: I think I got it eventually: we should only use f2 for testing its signature. If you call it you will pollute the temporary cache directory. The fact that it is a fixture makes it even more complicated to understand the comment I think ...

I may clean this up a bit after this PR is merged.

File: joblib/test/test_func_inspect.py (hunk 2 above)
Comment (karandesai-96) [inline joblib/test/test_func_inspect.py]: > When you notice unrelated shortcomings/problems while working on a PR, it is a good habit to change them in separate PRs.

I agree with this fact.
Okay, I'll make it `cached_func`.

**Comment by lesteve** [comment]
LGTM, merging, thanks a lot!

**Comment by karandesai-96** [comment]
Okay I'm leaving a long comment on the issue thread, please refer to it :smile:
